src/garage/tf/algos/rl2.py

```diff
@@ -1,15 +1,340 @@
 """Module for RL2.
 
-This module contains RL2Worker and the environment wrapper for RL2.
+This module contains RL2, RL2Worker and the environment wrapper for RL2.
 """
+import collections
+
 import akro
+from dowel import logger, tabular
 import gym
 import numpy as np
 
+import garage
+from garage import log_performance
+from garage import TrajectoryBatch
 from garage.envs import EnvSpec
+from garage.misc import tensor_utils as np_tensor_utils
+from garage.np.algos import MetaRLAlgorithm
 from garage.sampler import DefaultWorker
 
 
+class RL2(MetaRLAlgorithm):
+    """RL^2.
+
+    Reference: https://arxiv.org/pdf/1611.02779.pdf.
+
+    Args:
+        inner_algo (garage.np.algos.RLAlgorithm): Inner algorithm.
+        max_path_length (int): Maximum length for trajectories with respect
+            to RL^2. Notice that it is differen from the maximum path length
+            for the inner algorithm.
+        meta_batch_size (int): Meta batch size.
+        task_sampler (garage.experiment.TaskSampler): Task sampler.
+
+    """
+
+    def __init__(self, *, inner_algo, max_path_length, meta_batch_size,
+                 task_sampler):
+        assert isinstance(inner_algo, garage.tf.algos.BatchPolopt)
+        self._inner_algo = inner_algo
+        self._max_path_length = max_path_length
+        self._env_spec = inner_algo.env_spec
+        self._flatten_input = inner_algo.flatten_input
+        self._policy = inner_algo.policy
+        self._discount = inner_algo.discount
+        self._meta_batch_size = meta_batch_size
+        self._task_sampler = task_sampler
+
+    def train(self, runner):
+        """Obtain samplers and start actual training for each epoch.
+
+        Args:
+            runner (LocalRunner): LocalRunner is passed to give algorithm
+                the access to runner.step_epochs(), which provides services
+                such as snapshotting and sampler control.
+
+        Returns:
+            float: The average return in last epoch.
+
+        """
+        last_return = None
+
+        for _ in runner.step_epochs():
+            runner.step_path = runner.obtain_samples(
+                runner.step_itr,
+                env_update=self._task_sampler.sample(self._meta_batch_size))
+            tabular.record('TotalEnvSteps', runner.total_env_steps)
```

Comment:
Can we just stick this in `LocalRunner.step_epochs` while we're at it? I'm tired of finding an algorithm that forgot it, and now the X axis is silently something else.

Reply:
SGTM

Reply:
fixed